lib/api/api-fetch/headers.js

```diff
@@ -90,108 +100,193 @@ function validateArgumentLength (found, expected) {
+    if (this[index] === normalizedName) {
+      this[index + 1] += `, ${normalizedValue}`
     } else {
-      this[kHeadersList].splice(index, 0, normalizedName, normalizedValue)
+      this.splice(index, 0, normalizedName, normalizedValue)
     }
   }
 
   delete (...args) {
-    if (!isHeaders(this)) {
-      throw new InvalidThisError('Header')
-    }
-
     validateArgumentLength(args.length, 1)
 
     const [name] = args
 
     const normalizedName = normalizeAndValidateHeaderName(name)
 
-    const index = binarySearch(this[kHeadersList], normalizedName)
+    const index = binarySearch(this, normalizedName)
 
-    if (this[kHeadersList][index] === normalizedName) {
-      this[kHeadersList].splice(index, 2)
+    if (this[index] === normalizedName) {
+      this.splice(index, 2)
     }
   }
 
   get (...args) {
-    if (!isHeaders(this)) {
-      throw new InvalidThisError('Header')
-    }
-
     validateArgumentLength(args.length, 1)
 
     const [name] = args
 
     const normalizedName = normalizeAndValidateHeaderName(name)
 
-    const index = binarySearch(this[kHeadersList], normalizedName)
+    const index = binarySearch(this, normalizedName)
 
-    if (this[kHeadersList][index] === normalizedName) {
-      return this[kHeadersList][index + 1]
+    if (this[index] === normalizedName) {
+      return this[index + 1]
     }
 
     return null
   }
 
   has (...args) {
-    if (!isHeaders(this)) {
-      throw new InvalidThisError('Header')
-    }
-
     validateArgumentLength(args.length, 1)
 
     const [name] = args
 
     const normalizedName = normalizeAndValidateHeaderName(name)
 
-    const index = binarySearch(this[kHeadersList], normalizedName)
+    const index = binarySearch(this, normalizedName)
 
-    return this[kHeadersList][index] === normalizedName
+    return this[index] === normalizedName
   }
 
   set (...args) {
-    if (!isHeaders(this)) {
-      throw new InvalidThisError('Header')
-    }
-
     validateArgumentLength(args.length, 2)
 
     const [name, value] = args
 
     const normalizedName = normalizeAndValidateHeaderName(name)
     const normalizedValue = normalizeAndValidateHeaderValue(name, value)
 
-    const index = binarySearch(this[kHeadersList], normalizedName)
-    if (this[kHeadersList][index] === normalizedName) {
-      this[kHeadersList][index + 1] = normalizedValue
+    const index = binarySearch(this, normalizedName)
+    if (this[index] === normalizedName) {
+      this[index + 1] = normalizedValue
     } else {
-      this[kHeadersList].splice(index, 0, normalizedName, normalizedValue)
+      this.splice(index, 0, normalizedName, normalizedValue)
+    }
+  }
+}
+
+class Headers {
+  constructor (init = {}) {
+    // validateObject allowArray = true
+    if (!Array.isArray(init) && typeof init !== 'object') {
+      throw new TypeError('The argument \'init\' must be one of type Object or Array')
+    }
+    this[kHeadersList] = new HeadersList()
+    this[kGuard] = 'none'
+    fill(this, init)
```

Comment:
im not sure where I saw you ask this question but if im not mistaken you asked about why do we have the "fill" function, and it is because in the Request and Response classes when a Headers instance needs to be cloned, the algorithms do this weird thing where it will initialize an empty Headers instance and then run `fill` on it with the existing instance HeaderList. Might be avoidable if we write the Request/Response in a certain way

Reply:
I think I already solved this. PTAL at the response/request classes.